tactics/tactics.ml

```diff
@@ -889,23 +940,33 @@ let reduce redexp cl =
   in
   Proofview.Trace.name_tactic trace begin
   Proofview.Goal.enter begin fun gl ->
+  let env = Proofview.Goal.env gl in
   let hyps = concrete_clause_of (fun () -> Tacmach.New.pf_ids_of_hyps gl) cl in
   let nbcl = (if cl.concl_occs = NoOccurrences then 0 else 1) + List.length hyps in
   let check = match redexp with Fold _ | Pattern _ -> true | _ -> false in
```

Comment:
What does this one mean? Is it also incorrect?

Reply:
If you're talking about the `check` flag being true for `fold` and `pattern`, this is due to the fact that these "conversions" are not always legal. For instance, it might be ill-typed to pattern on a term whose transparent value is required for a term to be well-typed.

Reply:
Yes but this is assuming that external conversions do not have this problem, which nothing ensures, right?

Reply:
@ppedrot although for `fold` I don't even see the example where checking is required.

Reply:
> Yes but this is assuming that external conversions do not have this problem, which nothing ensures, right?

Once again, this is preserving the old behaviour.

> although for fold I don't even see the example where checking is required

You need type checking because *universes*. If you apply an anti-reduction, you might need to add new universe constraints, as in general reducing a term makes you lose constraints. There was an example with `simpl` refolding generating ill-typed terms at some point, but it was fixed.

Reply:
> Once again, this is preserving the old behaviour.

Sure, sorry, I mean this is something you should mention in the issue you are opening, it is ok to keep it as it is in this PR.

Reply:
Fact is, I don't even know how to define a custom reduction in a way that makes it usable through tactics. The only grammar that gets extended is the `eval ... in ...`, which does not use the same code path. The only way to hit this bug is to define a native reduction in ML, which doesn't happen often (if ever).

Reply:
>  I don't even know how to define a custom reduction in a way that makes it usable through tactics

The plugin [maple-mode](https://github.com/coq-contribs/maple-mode) defines custom reduction, combining declarations such as
```
Redexpr.declare_reduction "raw_simplify" (maple "simplify")
```
at the ml level and declarations such as
```
Ltac maple_simplify x := (eval raw_simplify in x).
Tactic Notation "simplify" ne_constr_list(l) := tac_iter (prove_rw maple_simplify) l.
```
at the vernac level.

Reply:
I was aware of this plugin, but this misses the point: the Maple plugin only uses the newly declared reduction via `eval ... in ...`. To trigger the problem discussed above, we would need a plugin that *both* declares an ML reduction *and* calls it via a `convert_hyp`-related ML function, as there is no way to trigger this code path via Coq code. This is very unlikely, because in practice one would rather call a lower level conversion function instead of going through the indirection of the named ML reduction.

Reply:
OK, believing you.